--- ../src-base/minecraft/net/minecraft/util/EntityDamageSourceIndirect.java
+++ ../src-work/minecraft/net/minecraft/util/EntityDamageSourceIndirect.java
@@ -38,4 +38,9 @@
         String s1 = s + ".item";
         return !itemstack.func_190926_b() && itemstack.func_82837_s() && I18n.func_94522_b(s1) ? new TextComponentTranslation(s1, new Object[] {p_151519_1_.func_145748_c_(), itextcomponent, itemstack.func_151000_E()}) : new TextComponentTranslation(s, new Object[] {p_151519_1_.func_145748_c_(), itextcomponent});
     }
+
+    @Nullable
+    public Entity getProximateDamageSource() {
+        return super.func_76346_g();
+    }
\ No newline at end of file
 }
