--- ../src-base/minecraft/net/minecraft/util/SoundEvent.java
+++ ../src-work/minecraft/net/minecraft/util/SoundEvent.java
@@ -1,12 +1,15 @@
 package net.minecraft.util;
 
 import net.minecraft.util.registry.RegistryNamespaced;
+import net.minecraftforge.fml.relauncher.ReflectionHelper;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.magmafoundation.magma.remapper.remappers.ReflectionMethodRemapper;
+import org.magmafoundation.magma.remapper.remappers.ReflectionRemapper;
 
-public class SoundEvent
+public class SoundEvent extends net.minecraftforge.registries.IForgeRegistryEntry.Impl<SoundEvent>
 {
-    public static final RegistryNamespaced<ResourceLocation, SoundEvent> field_187505_a = new RegistryNamespaced<ResourceLocation, SoundEvent>();
+    public static final RegistryNamespaced<ResourceLocation, SoundEvent> field_187505_a = net.minecraftforge.registries.GameData.getWrapper(SoundEvent.class);
     private final ResourceLocation field_187506_b;
     private static int field_187507_c;
 
@@ -21,6 +24,14 @@
         return this.field_187506_b;
     }
 
+    @Override
+    public String toString() {
+        return "SoundEvent{" +
+                "soundName=" + field_187506_b +
+                "soundEventId=" + field_187507_c +
+                '}';
+    }
+
\ No newline at end of file
     public static void func_187504_b()
     {
         func_187502_a("ambient.cave");
