--- ../src-base/minecraft/net/minecraft/world/chunk/storage/RegionFile.java
+++ ../src-work/minecraft/net/minecraft/world/chunk/storage/RegionFile.java
@@ -1,26 +1,28 @@
 package net.minecraft.world.chunk.storage;
 
 import com.google.common.collect.Lists;
-import java.io.BufferedInputStream;
-import java.io.BufferedOutputStream;
-import java.io.ByteArrayInputStream;
-import java.io.ByteArrayOutputStream;
-import java.io.DataInputStream;
-import java.io.DataOutputStream;
-import java.io.File;
-import java.io.IOException;
-import java.io.RandomAccessFile;
+import java.io.*;
+import java.nio.ByteBuffer;
+import java.nio.IntBuffer;
 import java.util.List;
 import java.util.zip.DeflaterOutputStream;
 import java.util.zip.GZIPInputStream;
 import java.util.zip.InflaterInputStream;
 import javax.annotation.Nullable;
+import net.minecraft.nbt.CompressedStreamTools;
+import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.server.MinecraftServer;
 
 public class RegionFile
 {
+    // Spigot start
+    // Minecraft is limited to 256 sections per chunk. So 1MB. This can easily be overriden.
+    // So we extend this to use the REAL size when the count is maxed by seeking to that section and reading the length.
+    private static final boolean ENABLE_EXTENDED_SAVE = Boolean.parseBoolean(System.getProperty("net.minecraft.world.chunk.storage.RegionFile.enableExtendedSave", "true"));
+    // Spigot end
     private static final byte[] field_76720_a = new byte[4096];
     private final File field_76718_b;
+    private File getFile() { return field_76718_b; }
     private RandomAccessFile field_76719_c;
     private final int[] field_76716_d = new int[1024];
     private final int[] field_76717_e = new int[1024];
@@ -42,7 +44,7 @@
 
             this.field_76719_c = new RandomAccessFile(p_i2001_1_, "rw");
 
-            if (this.field_76719_c.length() < 4096L)
+            if (this.field_76719_c.length() < 8192L) // Paper - headers should be 8192
             {
                 this.field_76719_c.write(field_76720_a);
                 this.field_76719_c.write(field_76720_a);
@@ -69,24 +71,54 @@
             this.field_76714_f.set(1, Boolean.valueOf(false));
             this.field_76719_c.seek(0L);
 
+            int k;
+
+            // Paper Start
+            ByteBuffer header = ByteBuffer.allocate(8192);
+            while (header.hasRemaining())  {
+                if (this.field_76719_c.getChannel().read(header) == -1) {
+                    throw new EOFException();
+                }
+            }
+            header.clear();
+            IntBuffer headerAsInts = header.asIntBuffer();
+            initOversizedState();
+            // Paper End
             for (int j1 = 0; j1 < 1024; ++j1)
             {
-                int k = this.field_76719_c.readInt();
+                k = headerAsInts.get(); // Paper
                 this.field_76716_d[j1] = k;
-
-                if (k != 0 && (k >> 8) + (k & 255) <= this.field_76714_f.size())
+                // Spigot start
+                int length = k & 255;
+                if (length == 255)
                 {
-                    for (int l = 0; l < (k & 255); ++l)
+                    if ((k >> 8) <= this.field_76714_f.size()) {
+                        // We're maxed out, so we need to read the proper length from the section
+                        this.field_76719_c.seek((k >> 8) * 4096);
+                        length = (this.field_76719_c.readInt() + 4)/ 4096 + 1;
+                        this.field_76719_c.seek(j1 * 4 + 4); //Go back to where we were
+                    }
+                }
+                if (k > 0 && (k >> 8) > 1 && (k >> 8) + (length) <= this.field_76714_f.size()) // Paper >= 1 as 0/1 are the headers, and negative isnt valid
+                {
+                    for (int l = 0; l < (length); ++l)
                     {
+                        // Spigot end
                         this.field_76714_f.set((k >> 8) + l, Boolean.valueOf(false));
                     }
                 }
+                // Spigot start
+                else if (k != 0) { // Paper
+                    org.bukkit.Bukkit.getLogger().log(java.util.logging.Level.WARNING, "Invalid chunk: ({0}, {1}) Offset: {2} Length: {3} runs off end file. {4}", new Object[]{j1 % 32, (int) (j1 / 32), k >> 8, length, p_i2001_1_});
+                    deleteChunk(j1); // Paper
+                }
+                // Spigot end
             }
 
             for (int k1 = 0; k1 < 1024; ++k1)
             {
-                int l1 = this.field_76719_c.readInt();
-                this.field_76717_e[k1] = l1;
+                k = headerAsInts.get(); // Paper
+                if (field_76716_d[k1] != 0) this.field_76717_e[k1] = k; // Paper - don't set timestamp if it got 0'd above due to corruption
             }
         }
         catch (IOException ioexception)
@@ -95,6 +127,12 @@
         }
     }
 
+    @Deprecated // TODO: remove (1.13)
+    public synchronized boolean chunkExists(int x, int z)
+    {
+        return func_76709_c(x, z);
+    }
+
     @Nullable
 
     public synchronized DataInputStream func_76704_a(int p_76704_1_, int p_76704_2_)
@@ -117,6 +155,13 @@
                 {
                     int j = i >> 8;
                     int k = i & 255;
+                    // Spigot start
+                    if (k == 255)
+                    {
+                        this.field_76719_c.seek(j * 4096);
+                        k = (this.field_76719_c.readInt() + 4) / 4096 + 1;
+                    }
+                    // Spigot end
 
                     if (j + k > this.field_76714_f.size())
                     {
@@ -169,7 +214,7 @@
     @Nullable
     public DataOutputStream func_76710_b(int p_76710_1_, int p_76710_2_)
     {
-        return this.func_76705_d(p_76710_1_, p_76710_2_) ? null : new DataOutputStream(new BufferedOutputStream(new DeflaterOutputStream(new RegionFile.ChunkBuffer(p_76710_1_, p_76710_2_))));
+        return this.func_76705_d(p_76710_1_, p_76710_2_) ? null : new DataOutputStream(new ChunkBuffer(p_76710_1_, p_76710_2_)); // Paper - remove middleware, move deflate to .close() for dynamic levels
     }
 
     protected synchronized void func_76706_a(int p_76706_1_, int p_76706_2_, byte[] p_76706_3_, int p_76706_4_)
@@ -179,11 +224,21 @@
             int i = this.func_76707_e(p_76706_1_, p_76706_2_);
             int j = i >> 8;
             int k = i & 255;
+            // Spigot start
+            if (k == 255)
+            {
+                this.field_76719_c.seek(j * 4096);
+                k = (this.field_76719_c.readInt() + 4) / 4096 + 1;
+            }
+            // Spigot end
             int l = (p_76706_4_ + 5) / 4096 + 1;
 
             if (l >= 256)
             {
-                return;
+                // Spigot start
+                if (!ENABLE_EXTENDED_SAVE) return;
+                org.bukkit.Bukkit.getLogger().log(java.util.logging.Level.WARNING,"Large Chunk Detected: ({0}, {1}) Size: {2} {3}", new Object[]{p_76706_1_, p_76706_2_, l, this.field_76718_b});
+                // Spigot end
             }
 
             if (j != 0 && k == l)
@@ -231,7 +286,7 @@
                 if (j1 >= l)
                 {
                     j = l1;
-                    this.func_76711_a(p_76706_1_, p_76706_2_, l1 << 8 | l);
+                    this.func_76711_a(p_76706_1_, p_76706_2_, l1 << 8 | (l > 255 ? 255 : l)); // Spigot
 
                     for (int j2 = 0; j2 < l; ++j2)
                     {
@@ -253,7 +308,7 @@
 
                     this.field_76715_g += 4096 * l;
                     this.func_76712_a(j, p_76706_3_, p_76706_4_);
-                    this.func_76711_a(p_76706_1_, p_76706_2_, j << 8 | l);
+                    this.func_76711_a(p_76706_1_, p_76706_2_, j << 8 | (l > 255 ? 255 : l)); // Spigot
                 }
             }
 
@@ -261,7 +316,7 @@
         }
         catch (IOException ioexception)
         {
-            ioexception.printStackTrace();
+            org.spigotmc.SneakyThrow.sneaky(ioexception); // Paper - we want the upper try/catch to retry this
         }
     }
 
@@ -310,6 +365,159 @@
         }
     }
 
+    // Paper start
+    public synchronized void deleteChunk(int j1) {
+        backup();
+        int k = field_76716_d[j1];
+        int x = j1 & 1024;
+        int z = j1 >> 2;
+        int offset = (k >> 8);
+        int len = (k & 255);
+        org.apache.logging.log4j.Logger logger = org.apache.logging.log4j.LogManager.getLogger();
+        String debug = "idx:" + + j1 + " - " + x + "," + z + " - offset: " + offset + " - len: " + len;
+        try {
+            RandomAccessFile file = field_76719_c;
+            file.seek(j1 * 4);
+            file.writeInt(0);
+            // clear the timestamp
+            file.seek(4096 + j1 * 4);
+            file.writeInt(0);
+            field_76717_e[j1] = 0;
+            field_76716_d[j1] = 0;
+            logger.error("Deleted corrupt chunk (" + debug + ") " + getFile().getAbsolutePath());
+        } catch (IOException e) {
+
+            logger.error("Error deleting corrupt chunk (" + debug + ") " + getFile().getAbsolutePath(), e);
+        }
+    }
+    private boolean backedUp = false;
+    private synchronized void backup() {
+        if (backedUp) {
+            return;
+        }
+        backedUp = true;
+        File file = this.getFile();
+        java.text.DateFormat formatter = new java.text.SimpleDateFormat("yyyy-MM-dd");
+        java.util.Date today = new java.util.Date();
+        File corrupt = new File(file.getParentFile(), file.getName() + "." + formatter.format(today) + ".corrupt");
+        if (corrupt.exists()) {
+            return;
+        }
+        org.apache.logging.log4j.Logger logger = org.apache.logging.log4j.LogManager.getLogger();
+        logger.error("Region file " + file.getAbsolutePath() + " was corrupt. Backing up to " + corrupt.getAbsolutePath() + " and repairing");
+        try {
+            java.nio.file.Files.copy(file.toPath(), corrupt.toPath());
+
+        } catch (IOException e) {
+            logger.error("Error backing up corrupt file" + file.getAbsolutePath(), e);
+        }
+    }
+
+    private final byte[] oversized = new byte[1024];
+    private int oversizedCount = 0;
+
+    private synchronized void initOversizedState() throws IOException {
+        File metaFile = getOversizedMetaFile();
+        if (metaFile.exists()) {
+            final byte[] read = java.nio.file.Files.readAllBytes(metaFile.toPath());
+            System.arraycopy(read, 0, oversized, 0, oversized.length);
+            for (byte temp : oversized) {
+                oversizedCount += temp;
+            }
+        }
+    }
+
+    private static int getChunkIndex(int x, int z) {
+        return (x & 31) + (z & 31) * 32;
+    }
+    synchronized boolean isOversized(int x, int z) {
+        return this.oversized[getChunkIndex(x, z)] == 1;
+    }
+    synchronized void setOversized(int x, int z, boolean oversized) throws IOException {
+        final int offset = getChunkIndex(x, z);
+        boolean previous = this.oversized[offset] == 1;
+        this.oversized[offset] = (byte) (oversized ? 1 : 0);
+        if (!previous && oversized) {
+            oversizedCount++;
+        } else if (!oversized && previous) {
+            oversizedCount--;
+        }
+        if (previous && !oversized) {
+            File oversizedFile = getOversizedFile(x, z);
+            if (oversizedFile.exists()) {
+                oversizedFile.delete();
+            }
+        }
+        if (oversizedCount > 0) {
+            if (previous != oversized) {
+                writeOversizedMeta();
+            }
+        } else if (previous) {
+            File oversizedMetaFile = getOversizedMetaFile();
+            if (oversizedMetaFile.exists()) {
+                oversizedMetaFile.delete();
+            }
+        }
+    }
+
+    private void writeOversizedMeta() throws IOException {
+        java.nio.file.Files.write(getOversizedMetaFile().toPath(), oversized);
+    }
+
+    private File getOversizedMetaFile() {
+        return new File(getFile().getParentFile(), getFile().getName().replaceAll("\\.mca$", "") + ".oversized.nbt");
+    }
+
+    private File getOversizedFile(int x, int z) {
+        return new File(this.getFile().getParentFile(), this.getFile().getName().replaceAll("\\.mca$", "") + "_oversized_" + x + "_" + z + ".nbt");
+    }
+
+    void writeOversizedData(int x, int z, NBTTagCompound oversizedData) throws IOException {
+        File file = getOversizedFile(x, z);
+        try (DataOutputStream out = new DataOutputStream(new BufferedOutputStream(new DeflaterOutputStream(new java.io.FileOutputStream(file), new java.util.zip.Deflater(java.util.zip.Deflater.BEST_COMPRESSION), 32 * 1024), 32 * 1024))) {
+            CompressedStreamTools.func_74800_a(oversizedData, out);
+        }
+        this.setOversized(x, z, true);
+
+    }
+
+    synchronized NBTTagCompound getOversizedData(int x, int z) throws IOException {
+        File file = getOversizedFile(x, z);
+        try (DataInputStream out = new DataInputStream(new BufferedInputStream(new InflaterInputStream(new java.io.FileInputStream(file))))) {
+            return CompressedStreamTools.func_74794_a(out);
+        }
+
+    }
+
+    private static final boolean USE_SPIGOT_OVERSIZED_METHOD = Boolean.getBoolean("Paper.useSpigotExtendedSaveMethod"); // Paper
+    static {
+        if (USE_SPIGOT_OVERSIZED_METHOD) {
+            org.bukkit.Bukkit.getLogger().log(java.util.logging.Level.SEVERE, "====================================");
+            org.bukkit.Bukkit.getLogger().log(java.util.logging.Level.SEVERE, "Using Spigot Oversized Chunk save method. Warning this will result in extremely fragmented chunks, as well as making the entire region file unable to be to used in any other software but Forge or Spigot (not usable in Vanilla or CraftBukkit). Paper's method is highly recommended.");
+            org.bukkit.Bukkit.getLogger().log(java.util.logging.Level.SEVERE, "====================================");
+        }
+    }
+
+    public class ChunkTooLargeException extends RuntimeException {
+        public ChunkTooLargeException(int x, int z, int sectors) {
+            super("Chunk " + x + "," + z + " of " + getFile().toString() + " is too large (" + sectors + "/256)");
+        }
+    }
+    private static class DirectByteArrayOutputStream extends ByteArrayOutputStream {
+        public DirectByteArrayOutputStream() {
+            super();
+        }
+
+        public DirectByteArrayOutputStream(int size) {
+            super(size);
+        }
+
+        public byte[] getBuffer() {
+            return this.buf;
+        }
+    }
+    // Paper end
+
     class ChunkBuffer extends ByteArrayOutputStream
     {
         private final int field_76722_b;
@@ -324,7 +532,34 @@
 
         public void close() throws IOException
         {
-            RegionFile.this.func_76706_a(this.field_76722_b, this.field_76723_c, this.buf, this.count);
+            // Paper start - apply dynamic compression
+            int origLength = this.count;
+            byte[] buf = this.buf;
+            DirectByteArrayOutputStream out = compressData(buf, origLength);
+            byte[] bytes = out.getBuffer();
+            int length = out.size();
+
+            RegionFile.this.func_76706_a(this.field_76722_b, this.field_76723_c, bytes, length);
         }
+        }
+
+    private static final byte[] compressionBuffer = new byte[1024 * 64]; // 64k fits most standard chunks input size even, ideally 1 pass through zlib
+    private static final java.util.zip.Deflater deflater = new java.util.zip.Deflater();
+    // since file IO is single threaded, no benefit to using per-region file buffers/synchronization, we can change that later if it becomes viable.
+    private static DirectByteArrayOutputStream compressData(byte[] buf, int length) throws IOException {
+        synchronized (deflater) {
+            deflater.setInput(buf, 0, length);
+            deflater.finish();
+
+
+            DirectByteArrayOutputStream out = new DirectByteArrayOutputStream(length);
+            while (!deflater.finished()) {
+                out.write(compressionBuffer, 0, deflater.deflate(compressionBuffer));
+            }
+            out.close();
+            deflater.reset();
+            return out;
     }
 }
+    // Paper end
+}
\ No newline at end of file
