--- ../src-base/minecraft/net/minecraft/nbt/NBTTagLong.java
+++ ../src-work/minecraft/net/minecraft/nbt/NBTTagLong.java
@@ -17,7 +17,7 @@
         this.field_74753_a = p_i45134_1_;
     }
 
-    void func_74734_a(DataOutput p_74734_1_) throws IOException
+    public void func_74734_a(DataOutput p_74734_1_) throws IOException
\ No newline at end of file
     {
         p_74734_1_.writeLong(this.field_74753_a);
     }
