--- ../src-base/minecraft/net/minecraft/nbt/NBTTagByte.java
+++ ../src-work/minecraft/net/minecraft/nbt/NBTTagByte.java
@@ -17,7 +17,7 @@
         this.field_74756_a = p_i45129_1_;
     }
 
-    void func_74734_a(DataOutput p_74734_1_) throws IOException
+    public void func_74734_a(DataOutput p_74734_1_) throws IOException
\ No newline at end of file
     {
         p_74734_1_.writeByte(this.field_74756_a);
     }
