--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityChest.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityChest.java
@@ -5,6 +5,7 @@
 import net.minecraft.block.BlockChest;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.entity.player.InventoryPlayer;
+import net.minecraft.init.Blocks;
 import net.minecraft.init.SoundEvents;
 import net.minecraft.inventory.Container;
 import net.minecraft.inventory.ContainerChest;
@@ -23,6 +24,11 @@
 import net.minecraft.util.math.AxisAlignedBB;
 import net.minecraft.util.math.BlockPos;
 
+import java.util.List;
+import org.bukkit.craftbukkit.v1_12_R1.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory;
+
 public class TileEntityChest extends TileEntityLockableLoot implements ITickable
 {
     private NonNullList<ItemStack> field_145985_p = NonNullList.<ItemStack>func_191197_a(27, ItemStack.field_190927_a);
@@ -37,6 +43,8 @@
     private int field_145983_q;
     private BlockChest.Type field_145982_r;
 
+    private int maxStack = MAX_STACK;
+
     public TileEntityChest()
     {
     }
@@ -45,7 +53,27 @@
     {
         this.field_145982_r = p_i46677_1_;
     }
+    public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
 
+    public void onOpen(CraftHumanEntity who) {
+        transaction.add(who);
+    }
+
+    public void onClose(CraftHumanEntity who) {
+        transaction.remove(who);
+    }
+
+    public List<HumanEntity> getViewers() {
+        return transaction;
+    }
+    public List<ItemStack> getContents() {
+        return this.field_145985_p;
+    }
+
+    public void setMaxStackSize(int size) {
+        maxStack = size;
+    }
+
     public int func_70302_i_()
     {
         return 27;
@@ -109,13 +137,14 @@
 
     public int func_70297_j_()
     {
-        return 64;
+        return maxStack;
     }
 
     public void func_145836_u()
     {
         super.func_145836_u();
         this.field_145984_a = false;
+        doubleChestHandler = null;
     }
 
     @SuppressWarnings("incomplete-switch")
@@ -167,6 +196,7 @@
     {
         if (!this.field_145984_a)
         {
+            if (this.field_145850_b == null || !this.field_145850_b.func_175697_a(this.field_174879_c, 1)) return; // Forge: prevent loading unloaded chunks when checking neighbors
             this.field_145984_a = true;
             this.field_145991_k = this.func_174911_a(EnumFacing.WEST);
             this.field_145990_j = this.func_174911_a(EnumFacing.EAST);
@@ -323,8 +353,18 @@
                 this.field_145987_o = 0;
             }
 
+            int oldPower = Math.max(0, Math.min(15, this.field_145987_o)); // CraftBukkit - Get power before new viewer is added
+
             ++this.field_145987_o;
+            if (this.field_145850_b == null) return;
             this.field_145850_b.func_175641_c(this.field_174879_c, this.func_145838_q(), 1, this.field_145987_o);
+            if (this.func_145838_q() == Blocks.field_150447_bR) {
+                int newPower = Math.max(0, Math.min(15, this.field_145987_o));
+
+                if (oldPower != newPower) {
+                    CraftEventFactory.callRedstoneChange(field_145850_b, field_174879_c.func_177958_n(), field_174879_c.func_177956_o(), field_174879_c.func_177952_p(), oldPower, newPower);
+                }
+            }
             this.field_145850_b.func_175685_c(this.field_174879_c, this.func_145838_q(), false);
 
             if (this.func_145980_j() == BlockChest.Type.TRAP)
@@ -338,17 +378,47 @@
     {
         if (!p_174886_1_.func_175149_v() && this.func_145838_q() instanceof BlockChest)
         {
+            int oldPower = Math.max(0, Math.min(15, this.field_145987_o)); // CraftBukkit - Get power before new viewer is added
             --this.field_145987_o;
             this.field_145850_b.func_175641_c(this.field_174879_c, this.func_145838_q(), 1, this.field_145987_o);
             this.field_145850_b.func_175685_c(this.field_174879_c, this.func_145838_q(), false);
 
+            // CraftBukkit start - Call redstone event
             if (this.func_145980_j() == BlockChest.Type.TRAP)
             {
+                int newPower = Math.max(0, Math.min(15, this.field_145987_o));
+                if (oldPower != newPower) {
+                    org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callRedstoneChange(field_145850_b, field_174879_c.func_177958_n(), field_174879_c.func_177956_o(), field_174879_c.func_177952_p(), oldPower, newPower);
+                }
                 this.field_145850_b.func_175685_c(this.field_174879_c.func_177977_b(), this.func_145838_q(), false);
             }
+            // CraftBukkit end
         }
     }
 
+    public net.minecraftforge.items.VanillaDoubleChestItemHandler doubleChestHandler;
+
+    @SuppressWarnings("unchecked")
+    @Override
+    @Nullable
+    public <T> T getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @Nullable EnumFacing facing)
+    {
+        if (capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY)
+        {
+            if(doubleChestHandler == null || doubleChestHandler.needsRefresh())
+                doubleChestHandler = net.minecraftforge.items.VanillaDoubleChestItemHandler.get(this);
+            if (doubleChestHandler != null && doubleChestHandler != net.minecraftforge.items.VanillaDoubleChestItemHandler.NO_ADJACENT_CHESTS_INSTANCE)
+                return (T) doubleChestHandler;
+        }
+        return super.getCapability(capability, facing);
+    }
+
+    public net.minecraftforge.items.IItemHandler getSingleChestHandler()
+    {
+        return super.getCapability(net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, null);
+    }
+
+
\ No newline at end of file
     public void func_145843_s()
     {
         super.func_145843_s();
