--- ../src-base/minecraft/net/minecraftforge/fluids/FluidUtil.java
+++ ../src-work/minecraft/net/minecraftforge/fluids/FluidUtil.java
@@ -49,6 +49,9 @@
 import net.minecraftforge.items.CapabilityItemHandler;
 import net.minecraftforge.items.IItemHandler;
 import net.minecraftforge.items.ItemHandlerHelper;
+import org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory;
+import org.bukkit.event.player.PlayerBucketEmptyEvent;
+import org.bukkit.event.player.PlayerBucketFillEvent;
 
 public class FluidUtil
 {
@@ -567,6 +570,17 @@
 
         if (block instanceof IFluidBlock || block instanceof BlockLiquid)
         {
+            // Mohist start - calling bukkit event before pickup custom fluid
+            if (playerIn != null) {
+                // Send default empty bucket
+                PlayerBucketFillEvent event = CraftEventFactory.callPlayerBucketFillEvent(playerIn, pos.getX(), pos.getY(), pos.getZ(), null, emptyContainer.copy(), Items.BUCKET);
+                // Changes from event not handled, only canceling
+                if (event.isCancelled()) {
+                    return FluidActionResult.FAILURE;
+                }
+            }
+            // Mohist end - calling bukkit event before pickup custom fluid
+
             IFluidHandler targetFluidHandler = getFluidHandler(worldIn, pos, side);
             if (targetFluidHandler != null)
             {
@@ -621,6 +635,16 @@
             return false;
         }
 
+        // Mohist start - calling bukkit event before place custom fluid
+        if (player != null) {
+            PlayerBucketEmptyEvent event = CraftEventFactory.callPlayerBucketEmptyEvent(player, pos.getX(), pos.getY(), pos.getZ(), null, player.getHeldItem(EnumHand.MAIN_HAND).copy());
+            // Changes from event not handled, only canceling
+            if (event.isCancelled()) {
+                return false;
+            }
+        }
+        // Mohist end - calling bukkit event before place custom fluid
+
         Fluid fluid = resource.getFluid();
         if (fluid == null || !fluid.canBePlacedInWorld())
         {
