--- ../src-base/minecraft/net/minecraftforge/fluids/UniversalBucket.java
+++ ../src-work/minecraft/net/minecraftforge/fluids/UniversalBucket.java
@@ -44,6 +44,8 @@
 import net.minecraftforge.fml.common.eventhandler.EventPriority;
 import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
 import net.minecraftforge.items.ItemHandlerHelper;
+import org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory;
+import org.bukkit.event.player.PlayerBucketFillEvent;
 
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
@@ -214,6 +216,25 @@
     {
         if (event.getResult() != Event.Result.DEFAULT)
         {
+            // Mohist start - checking when result is ALLOW
+            if (event.getResult() == Event.Result.ALLOW) {
+                if (event.getEntityPlayer() != null) {
+                    RayTraceResult target = event.getTarget();
+                    if (target != null) {
+                        BlockPos pos = target.getBlockPos();
+                        PlayerBucketFillEvent bEvent = CraftEventFactory.callPlayerBucketFillEvent(
+                                event.getEntityPlayer(), pos.getX(), pos.getY(), pos.getZ(), null,
+                                event.getEntityPlayer().getHeldItemMainhand().copy(), Items.BUCKET);
+                        // Changes from event not handled, only canceling
+                        if (bEvent.isCancelled()) {
+                            event.setCanceled(true);
+                            event.setResult(Event.Result.DENY);
+                        }
+                    }
+                }
+            }
+            // Mohist end - checking when result is ALLOW
+
             // event was already handled
             return;
         }
