--- ../src-base/minecraft/net/minecraft/entity/passive/EntityPig.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityPig.java
@@ -37,6 +37,7 @@
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.world.World;
 import net.minecraft.world.storage.loot.LootTableList;
+import org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory;
 
 public class EntityPig extends EntityAnimal
 {
@@ -188,7 +189,7 @@
 
     public void func_70645_a(DamageSource p_70645_1_)
     {
-        super.func_70645_a(p_70645_1_);
+        // super.onDeath(cause); // CraftBukkit - Moved to end
 
         if (!this.field_70170_p.field_72995_K)
         {
@@ -197,6 +198,7 @@
                 this.func_145779_a(Items.field_151141_av, 1);
             }
         }
+        super.func_70645_a(p_70645_1_); // CraftBukkit - Moved from above
     }
 
     @Nullable
@@ -222,11 +224,26 @@
         }
     }
 
-    public void func_70077_a(EntityLightningBolt p_70077_1_)
+    public void func_70077_a(@Nullable EntityLightningBolt p_70077_1_)
     {
         if (!this.field_70170_p.field_72995_K && !this.field_70128_L)
         {
+            if (p_70077_1_ == null)
+                p_70077_1_ = new EntityLightningBolt(this.field_70170_p, this.field_70165_t, this.field_70163_u, this.field_70161_v, false);
             EntityPigZombie entitypigzombie = new EntityPigZombie(this.field_70170_p);
+
+            // Paper start
+            if(CraftEventFactory.callEntityZapEvent(this, p_70077_1_, entitypigzombie).isCancelled()){
+                return;
+            }
+            // Paper end
+
+            // CraftBukkit start
+            if (CraftEventFactory.callPigZapEvent(this, p_70077_1_, entitypigzombie).isCancelled()) {
+                return;
+            }
+            // CraftBukkit end
+
             entitypigzombie.func_184201_a(EntityEquipmentSlot.MAINHAND, new ItemStack(Items.field_151010_B));
             entitypigzombie.func_70012_b(this.field_70165_t, this.field_70163_u, this.field_70161_v, this.field_70177_z, this.field_70125_A);
             entitypigzombie.func_94061_f(this.func_175446_cd());
@@ -237,6 +254,7 @@
                 entitypigzombie.func_174805_g(this.func_174833_aM());
             }
 
+            this.field_70170_p.prepareSpawnEntity(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING); // Magma - Added a reason for spawning this creature
\ No newline at end of file
             this.field_70170_p.func_72838_d(entitypigzombie);
             this.func_70106_y();
         }
